### PR TITLE
CRF decoder (and runner) for sequence labeling

### PR DESCRIPTION
**Pull request by varisd**
- Extended sequence_labeler with a child class that uses CRF classifier (from tf.contrib.crf) with trainable parameters on top of the decoded logits.
- Added separate runner that works with this decoder.
- Scores are only computed for the whole sequence.
- Does not support ensembling.

see tests/crf_labeler.ini for sample configuration

### CODE REVIEW

File: neuralmonkey/decoders/sequence_labeler.py [hunk 1]
```diff
@@ -94,7 +94,8 @@ def logits(self) -> tf.Tensor:
             embedded_inputs, dweights_4d, [1, 1, 1, 1], "SAME")
         dmultiplication_3d = tf.squeeze(dmultiplication, squeeze_dims=[2])
 
-        logits = multiplication_3d + dmultiplication_3d + biases_3d
+        logits = multiplication_3d + dmultiplication_3d
+        logits = logits + biases_3d
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: +=

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: to se nevejde na jednu řádku?

File: neuralmonkey/decoders/sequence_labeler.py [hunk 2]
```diff
@@ -139,3 +140,59 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
             fd[self.train_weights] = paddings.T
 
         return fd
+
+
+class CRFLabeler(SequenceLabeler):
+    """CRF Classifier assing a label to each encoder's state."""
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: ``encoder state`` znamená to samý

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: assing -> that assigns

File: neuralmonkey/decoders/sequence_labeler.py [hunk 3]
```diff
@@ -139,3 +140,59 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: Optional[str] = None,
+                 load_checkpoint: Optional[str] = None) -> None:
+        SequenceLabeler.__init__(self, name, encoder, vocabulary, data_id,
+                                 dropout_keep_prob, save_checkpoint,
+                                 load_checkpoint)
+
+        self.transition_params = tf.get_variable(
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: nepoužíváš scope tohohle model partu: obalit do ``with self.use_scope():`` nebo něco takovýho. Je to funkce na model partu

File: neuralmonkey/decoders/sequence_labeler.py [hunk 4]
```diff
@@ -139,3 +140,59 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+    """CRF Classifier assing a label to each encoder's state."""
+
+    def __init__(self,
+                 name: str,
+                 encoder: Union[RecurrentEncoder, SentenceEncoder],
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: Optional[str] = None,
+                 load_checkpoint: Optional[str] = None) -> None:
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: ``check_argument_types()``

File: neuralmonkey/decoders/sequence_labeler.py [hunk 5]
```diff
@@ -139,3 +140,59 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
+
+        self.transition_params = tf.get_variable(
+            "transition_params",
+            [len(self.vocabulary) - 3, len(self.vocabulary) - 3],
+            initializer=tf.glorot_uniform_initializer())
+
+    @tensor
+    def logprobs(self) -> tf.Tensor:
+        raise NotImplementedError("Logarithmic probabilities are not "
+                                  "supported by the CRFLabeler")
```
Comment (jindrahelcl) [inline neuralmonkey/decoders/sequence_labeler.py]: Tohle je divný. Musí bejt CRF labeler potomkem sequence labeleru? Nemůže to bejt spíš komponenta nad tim labelerem, která bude používat jeho logity? V konstruktoru si ten labeler může vytvořit a mít ho jako atribut, ne?

File: neuralmonkey/runners/crf_runner.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+            raise ValueError("CRFRunner does not support ensembling.")
+
+        loss = results[0].get("loss", 0.)
+        input_mask = results[0]["input_mask"]
+
+        decoded_indices = results[0]["decoded_sequence"]
+
+        assert (input_mask == results[0]["input_mask"]).all()
+
+        # CAUTION! FABULOUS HACK BELIEVE ME
```
Comment (jindrahelcl) [inline neuralmonkey/runners/crf_runner.py]: omg kolik je toho copy-paste? Nedá se z toho vyčlenit dekodovaci funkce, kterou budou sdilet oba runnery?

File: neuralmonkey/decoders/sequence_labeler.py (hunk 1 above)
Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Jo, je mozne, ze mi to tam zustalo po debugovani.

File: neuralmonkey/runners/crf_runner.py (hunk 6 above)
Comment (varisd) [reply neuralmonkey/runners/crf_runner.py]: Jo, to by slo. Teoreticky by to mohl byt ciste jen Label Runner s tim, ze ta CRF vrstva by byl jen parametr runneru/dekoderu.

File: neuralmonkey/decoders/sequence_labeler.py [hunk 7]
```diff
@@ -139,3 +140,59 @@ def feed_dict(self, dataset: Dataset, train: bool = False) -> FeedDict:
...
 
         return fd
+
+
+class CRFLabeler(SequenceLabeler):
+    """CRF Classifier assing a label to each encoder's state."""
+
+    def __init__(self,
+                 name: str,
+                 encoder: Union[RecurrentEncoder, SentenceEncoder],
```
Comment (jlibovicky) [inline neuralmonkey/decoders/sequence_labeler.py]: Nemůže to být něco obecnějšího? `TemporalStateful`?

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: No potrebujes objekt, u ktereho mas zarucene, ze ma nadefinovane logity (nebo obecneji vystupni matici ve formatu [batch_size, seq_len]). Myslim, ze TemporalStateful tuhle podminku nesplnuje.

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: temporal stateful to je, má to (batch, time, whatever) matici a masky. Nicméně distribuce přes slovník (ať už normalizovaný nebo ne) bych stavama spíš nenazýval.

pořád si myslim že to nemá dědit od sequence labeleru vůbec;  ta CRF vrstva mi přijde jen jako jinej způsob převodu logitů (batch, time, vocabulary) na symboly (batch, time) a loss (), kde default je argmax a sequence_loss. Nebo se pletu?

Comment (varisd) [reply neuralmonkey/decoders/sequence_labeler.py]: Jo, misto po jednotlivych symbolech (tj. bez ohledu na dekodovane sousedy) se tady ucis dekodovat celou sekvenci najednou.
Jeste je otazka, jestli by byl problem stavet CRF vrstvu nad normalizovanymi logity (tedy po softmaxu). Pokud by to slo, tak by se tam daly implementovat i ensembly.

Comment (jindrahelcl) [reply neuralmonkey/decoders/sequence_labeler.py]: Zamyslel bych se jak to abstrahovat, ale před tim bych teda nechal tohle pracovat s čímkoliv (`Any`) co má za runtimu atribut `logits`. Plus udělat spešl syntactic sugar verzi, která si udělá svuj sequence labeler, podobně jako si ``encoders.SentenceEncoder`` dělá svou ``EmbeddedSequence``.

**Comment by jindrahelcl** [comment]
tak co? něco novýho?

**Comment by varisd** [comment]
Hele, tedka se chci vrhnout na ten PR k transformeru, to ma vetsi prioritu. Toto mozna zpracuju, az kdyz budeme mit hotove experimenty na LREC.

**Comment by jindrahelcl** [comment]
já jen žes mě hnal ať ti udělám review

**Comment by jindrahelcl** [comment]
Tak co, už máš hotové experimenty na LREC.. :-)

**Comment by varisd** [comment]
Jo, jeste je potreba ten bastl trochu refaktorovat. Mam to na TODO listu.

A mozna bych ty veci z vetve lrec_mwe rozdelil do vice PR, aby to bylo prehlednejsi. Co myslite?

**Comment by jlibovicky** [comment]
Co tam všechno je? CRF a character-level word embeddings? To bych udělal jako dva samostatné PRs.

**Comment by varisd** [comment]
Jo, tohle a jeste separatni processor+evaluator pro ten MWE task

**Comment by jindrahelcl** [comment]
Jasně, klidně to rozděl.

Dne 8. 3. 2018 11:47 dop. napsal uživatel "Dušan Variš" <
notifications@github.com>:

Jo, jeste je potreba ten bastl trochu refaktorovat. Mam to na TODO listu.

A mozna bych ty veci z vetve lrec_mwe rozdelil do vice PR, aby to bylo
prehlednejsi. Co myslite?

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/ufal/neuralmonkey/pull/612#issuecomment-371448590>, or mute
the thread

**Comment by jindrahelcl** [comment]
Since the resolution four months ago was to divide the commits to two pull request, this one is no longer relevant. If you someday finish the work, open up a new pull request. In the meantime, please keep your TODO list on a different place.  :-)
